### ISSUE TITLE
Support event inheritance in the psharp syntax

### ISSUE THREAD

**Issue by akashlal** [p-org/PSharp#277]
To allow something like:
```
event e1 (a: string);
event e2: e1 (b: int);
```
Which should translate to:
```
public class e1 : Event
{
   public string a;
   public e1(string a) : base() { this.a = a; }
}

public class e2 : e1
{
   public int b;
   public e2(string a, int b) : base(a) { this.b = b; }
}
```

Note that there is some analysis required to correctly set up the constructor of `e2`.

**Comment by akashlal** [comment]
Perhaps the easier option is to translate it to the following:
```
public class e1 : Event
{
   public string a;
   public e1(string a) : base() { this.a = a; }
   public e1(e1 x) : base() { this.a = x.a; }
}

public class e2 : e1
{
   public string b;
   public e2(e1 x, int b) : base(x) { this.b = b; }
}

```

**Comment by akashlal** [comment]
Further:
```
send(target, e2, "hello", 10);
```

would become:

```
this.Send(target, new e2(new e1("hello"), 10));
```
There is a wasteful `e1` object that is created, but it avoids doing any analysis in the syntax rewriter (which would be impossible if `e1` was declared in another file).

@pdeligia @TedHartMS What do you guys think?

**Comment by TedHartMS** [comment]
Since e1 may be defined in another file, the only way I can think of to define e2 in terms of e1's parameters (rather than an instance of e1) is by repeating the parameters of e1, e.g.:
```
event e2: e1 (a:string) (b: int);
```
which allows the first rewrite and saves the run-time allocation but is awkward. (Since e2 might be in a different file the send() call can't do anything clever either).

**Comment by akashlal** [comment]
I like Ted's suggestion. I believe supporting `send` isn't important. The user can just use the `C#` syntax right there in the `P#` file. In fact, we can eventually deprecate the send rewriting. 

We can do the following: (1) support `event e2 : e1 (b: int)` when `e1` is declared in the same file. Otherwise we complain with "definition of `e1` not present in the same file" and ask the user to give us the payload of `e1` using Ted's syntax: `event e2: e1 (a:string) (b: int);`

In each case, we spit out the full constructor for the event (as we do currently), which is what the `send` will use.

**Comment by akashlal** [comment]
@TedHartMS Do you agree with my latest proposal? Can I assign this item to you please?

**Comment by TedHartMS** [comment]
Yes and yes

**Comment by akashlal** [comment]
Fixed via #283
